src/systems/audio-debug-system.js

```diff
@@ -0,0 +1,323 @@
+    } else if (distanceModel == INVERSE) {
+      attenuation = att_inverse(d, rolloffFactor, refDistance);
+    } else if (distanceModel == EXPONENTIAL) {
+      attenuation = att_exponential(d, rolloffFactor, refDistance);
+    }
+    attenuation = clamp(attenuation, 0.0, 1.0);
+
+    // Waves
+    float v = sin((d * 2.0 * kPi) - (time * 0.005)) + 1.0;
+    float waves = circle.a * attenuation * v * vNormal.y;
+    
+    // Output final color.
+    return vec4( circle.rgb, waves );
+  }
+
+  void main() {
+    // Draw background
+    vec4 background = vec4(1.0, 1.0, 1.0, 0.0);
+
+    for (int i=0; i<count; i++) {
+      vec2 center = sourcePosition[i].xz - vUv;
+
+      // Calculate distance to (0,0).
+      float d = length( center );
+
+      // Optional start Offset.
+      vec2 orientation = normalize(sourceOrientation[i].xz);
+      float ang = atan(-orientation.x, orientation.y) * kInvPi * 0.5;
+      // Rotate to start drawing facing front
+      ang -= 0.5;
+      float startOffset = mod(ang, 1.0);
+
+      // Draw inner cone
+      float innerAngle = coneInnerAngle[i] * kDegToRad * kInvPi * 0.5;
+      float innerStartAngle = startOffset + innerAngle * 0.5;
+      vec4 innerLayer = circle(center, d, innerAngle, 10000.0, 1.0, colorInner, innerStartAngle);
+      innerLayer = att(d, innerLayer, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, innerLayer, innerLayer.a);
+
+      // Draw outer cone
+      float outerAngle = coneOuterAngle[i] * kDegToRad * kInvPi * 0.5;
+      float outerAngleDiffHalf = (outerAngle - innerAngle) * 0.5;
+      vec4 outerLayer1 = circle(center, d, outerAngleDiffHalf, 10000.0, 1.0, colorOuter, innerStartAngle + outerAngleDiffHalf);
+      outerLayer1 = att(d, outerLayer1, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, outerLayer1, outerLayer1.a);
+      vec4 outerLayer2 = circle(center, d, outerAngleDiffHalf, 10000.0, 1.0, colorOuter, innerStartAngle - innerAngle);
+      outerLayer2 = att(d, outerLayer2, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, outerLayer2, outerLayer2.a * (clipped[i] ? 0.0 : 1.0));
+
+      // Draw base
+      vec4 baseLayer = circle(center, d, 1.0, 1.0, 0.1, vec3(0.5, 0.5, 0.5), 0.0);
+      background = mix(background, baseLayer, baseLayer.a);
+
+      // Draw gain
+      float g = clamp(gain[i], 0.0, 1.0);
+      vec4 gainLayer = circle(center, d, g, 1.0, 0.5, colorGain, startOffset);
+      background = mix(background, gainLayer, gainLayer.a);
+      if (gain[i] > 1.0) {
+        vec4 overGainLayer = circle(center, d, gain[i] - g, 1.0, 0.5, vec3(1.0, 0.0, 0.0), startOffset);
+        background = mix(background, overGainLayer, overGainLayer.a);
+      }
+    }
+    
+    // Blend
+    gl_FragColor = vec4(background.rgb, background.a * vNormal.y);
+  }
+`;
+
+AFRAME.registerSystem("audio-debug", {
```

Comment:
Awesome job on the visuals here, looks really slick, and this will definitely be super useful for both us tuning default audio settings as well as space designers for tuning their spaces. Bet it was pretty fun to write too :)

Reply:
Yeah, I'm pretty happy about how is looking, I wanted it to be a elegant, simple and clear and I think it's pretty ok. Definitely fun stuff.